File: 0015.3sum/3sum.cpp
class Solution {
public:
    //完全参考了https://leetcode.com/problems/3sum/discuss/7402/Share-my-AC-C++-solution-around-50ms-O(N*N)-with-explanation-and-comments。
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        if(nums.size()<3) return result;
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-1;i++){
            int target=-nums[i];
            int second=i+1;
            int third=nums.size()-1;
            //cout<<"1:"<<i<<" "<<second<<" "<<third<<endl;
            while(second<third){
                int sum=nums[second]+nums[third];
                if(sum<target) second++;
                else if(sum>target) third--;
                else{
                    vector<int> tmp{nums[i],nums[second],nums[third]};
                    result.push_back(tmp);
                    //cout<<"result:"<<nums[i]<<" "<<nums[second]<<" "<<nums[third]<<endl;
                    //cout<<"result_index:"<<i<<" "<<second<<" "<<third<<endl;
                    second++;
                    third--;
                    while(second<third && nums[second]==tmp[1]){
                        second++;
                        //cout<<"2:"<<i<<" "<<second<<" "<<third<<endl;
                    }
                    //cout<<"2':"<<i<<" "<<second<<" "<<third<<endl;
                    while(second<third && nums[third-1]==tmp[2]){
                        third--;
                        //cout<<"3:"<<i<<" "<<second<<" "<<third<<endl;
                    }
                    //cout<<"3':"<<i<<" "<<second<<" "<<third<<endl;
                }
            }
            while(i+1<nums.size() && nums[i]==nums[i+1]){
                i++;
                //注意这里完毕后，在for循环里面还会再+1，这样才避免了重复的问题。
                //cout<<"4:"<<i<<" "<<second<<" "<<third<<endl;
            }
        }
        
        
        return result;
    }
    
    
    //更优雅的解法！！！：https://leetcode.com/problems/3sum/discuss/7438/C++-two-pointes-solution-(easy-way-to-handle-duplicaiton).
};

//  0  1  2  3  4  5
// -4 -1 -1  0  1  2

File: 0015.3sum/3sum.py
class Solution(object):
    def threeSum1(self, nums):#改解法超时！！太暴力了
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result=[]
        dict={}
        for index in nums:
            dict.update({index:[]})
        for indexa in range(len(nums)):
            a=nums[indexa]
            for indexb in range(indexa+1,len(nums)):
                b=nums[indexb]
                if b in dict[a]:#剪枝
                    continue
                for indexc in range(indexb+1,len(nums)):
                    c=nums[indexc]
                    if c in dict[a] or c in dict[b]:
                        continue
                    if a+b+c==0 and (b not in dict[a] and c not in dict[a]) and (a not in dict[b] and c not in dict[b]) and (a not in dict[c] and b not in dict[a]):
                        dict[a].extend([b,c])
                        dict[b].extend([a,c])
                        dict[c].extend([b,a])
                        result.append([a,b,c])
        print (dict)
        return result
      
    #参考了Discuss
    #一点心得：还是双指针好使啊
    def threeSum(self, nums):
        nums.sort()
        result=[]
        for indexa in range(len(nums)):
            if indexa>0 and nums[indexa]==nums[indexa-1]:
                continue
            a=nums[indexa]
            indexb=min(indexa+1,len(nums)-1)
            indexc=len(nums)-1
            while indexb<indexc:
                if a+nums[indexb]+nums[indexc]<0:
                    indexb+=1
                elif a+nums[indexb]+nums[indexc]>0:
                    indexc-=1
                else :
                    result.append([a,nums[indexb],nums[indexc]])
                    while indexb<indexc and nums[indexb]==nums[indexb+1] :
                        indexb+=1
                    while indexb<indexc and nums[indexc]==nums[indexc-1] :
                        indexc-=1
                    indexb+=1
                    indexc-=1
        print (dict)
        return result
        
        
        
        
        
        
        
        
        
    
    